src/Storages/S3Queue/S3QueueMetadata.cpp

```diff
@@ -0,0 +1,505 @@
+    }
+
+    Strings failed_nodes;
+    code = zk_client->tryGetChildren(zookeeper_failed_path, failed_nodes);
+    if (code != Coordination::Error::ZOK)
+    {
+        if (code == Coordination::Error::ZNONODE)
+        {
+            LOG_TEST(log, "Path {} does not exist", zookeeper_failed_path.string());
+        }
+        else
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected error: {}", magic_enum::enum_name(code));
+    }
+
+    const size_t nodes_num = processed_nodes.size() + failed_nodes.size();
+    if (!nodes_num)
+    {
+        LOG_TEST(log, "There are neither processed nor failed nodes (in {} and in {})",
+                 zookeeper_processed_path.string(), zookeeper_failed_path.string());
+        return;
+    }
+
+    chassert(settings.s3queue_tracked_files_limit || settings.s3queue_tracked_file_ttl_sec);
+    const bool check_nodes_limit = settings.s3queue_tracked_files_limit > 0;
+    const bool check_nodes_ttl = settings.s3queue_tracked_file_ttl_sec > 0;
+
+    const bool nodes_limit_exceeded = nodes_num > settings.s3queue_tracked_files_limit;
+    if ((!nodes_limit_exceeded || !check_nodes_limit) && !check_nodes_ttl)
+    {
+        LOG_TEST(log, "No limit exceeded");
+        return;
+    }
+
+    LOG_TRACE(log, "Will check limits for {} nodes", nodes_num);
+
+    /// Create a lock so that with distributed processing
+    /// multiple nodes do not execute cleanup in parallel.
+    auto ephemeral_node = zkutil::EphemeralNodeHolder::tryCreate(zookeeper_cleanup_lock_path, *zk_client, toString(getCurrentTime()));
+    if (!ephemeral_node)
+    {
+        LOG_TEST(log, "Cleanup is already being executed by another node");
+        return;
+    }
+    /// TODO because of this lock we might not update local file statuses on time on one of the nodes.
+
+    struct Node
+    {
+        std::string zk_path;
+        S3QueueIFileMetadata::NodeMetadata metadata;
+    };
+    auto node_cmp = [](const Node & a, const Node & b)
+    {
+        return std::tie(a.metadata.last_processed_timestamp, a.metadata.file_path)
+            < std::tie(b.metadata.last_processed_timestamp, b.metadata.file_path);
+    };
+
+    /// Ordered in ascending order of timestamps.
+    std::set<Node, decltype(node_cmp)> sorted_nodes(node_cmp);
+
+    for (const auto & node : processed_nodes)
+    {
+        const std::string path = zookeeper_processed_path / node;
+        try
+        {
+            std::string metadata_str;
+            if (zk_client->tryGet(path, metadata_str))
+            {
+                sorted_nodes.emplace(path, S3QueueIFileMetadata::NodeMetadata::fromString(metadata_str));
+                LOG_TEST(log, "Fetched metadata for node {}", path);
+            }
+            else
+                LOG_ERROR(log, "Failed to fetch node metadata {}", path);
+        }
+        catch (const zkutil::KeeperException & e)
+        {
+            if (e.code != Coordination::Error::ZCONNECTIONLOSS)
```

Comment:
maybe you want to check that it's not any hardware error? (with `isHardwareError`)